File: test/TaskStepsTest.java
import org.junit.jupiter.api.*;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import play.test.*;
import services.TaskSteps;

import java.util.List;

/**
 * JCLAW-260: unit tests for {@link TaskSteps#parse} and
 * {@link TaskSteps#flattenForPrompt}. Pure functions — no DB needed.
 */
class TaskStepsTest extends UnitTest {

    // --- parse ---

    @Test
    void plainTextParsesToSingleStep() {
        assertEquals(List.of("Water the plants"), TaskSteps.parse("Water the plants"));
    }

    @Test
    void jsonArrayParsesToOrderedSteps() {
        var steps = TaskSteps.parse("[\"Fetch orders\",\"Summarise\",\"Post to Slack\"]");
        assertEquals(List.of("Fetch orders", "Summarise", "Post to Slack"), steps);
    }

    @Test
    void multiStepJsonRoundTrips() {
        // The LLM-supplied order is preserved exactly.
        assertEquals(List.of("a", "b", "c"), TaskSteps.parse("[\"a\",\"b\",\"c\"]"));
    }

    @Test
    void singleElementArrayParsesToOneStep() {
        assertEquals(List.of("only step"), TaskSteps.parse("[\"only step\"]"));
    }

    @Test
    void nullAndBlankParseToEmptyList() {
        assertTrue(TaskSteps.parse(null).isEmpty());
        assertTrue(TaskSteps.parse("").isEmpty());
        assertTrue(TaskSteps.parse("   ").isEmpty());
    }

    /**
     * Every input that isn't a JSON array of strings falls back to a
     * single-element list holding the raw text verbatim: malformed JSON,
     * an array with non-string elements, an empty array, and a JSON object.
     */
    @ParameterizedTest(name = "fallsBackToSingleStep[{0}]")
    @ValueSource(strings = {
            "[not valid json",
            "[\"ok\", 42, true]",
            "[]",
            "{\"foo\":\"bar\"}"
    })
    void nonStringArrayInputFallsBackToSingleStep(String raw) {
        assertEquals(List.of(raw), TaskSteps.parse(raw));
    }

    // --- flattenForPrompt ---

    @Test
    void flattenPlainTextIsVerbatim() {
        assertEquals("Water the plants", TaskSteps.flattenForPrompt("Water the plants"));
    }

    @Test
    void flattenMultiStepIsNumberedList() {
        var flat = TaskSteps.flattenForPrompt("[\"Fetch orders\",\"Summarise\",\"Post to Slack\"]");
        assertEquals("1. Fetch orders\n2. Summarise\n3. Post to Slack", flat);
    }

    @Test
    void flattenSingleElementArrayEqualsPlainString() {
        // No-regression invariant: a one-element array and the equivalent
        // plain string flatten to byte-identical prompts (no numbering).
        assertEquals("do X", TaskSteps.flattenForPrompt("[\"do X\"]"));
        assertEquals(TaskSteps.flattenForPrompt("do X"),
                TaskSteps.flattenForPrompt("[\"do X\"]"));
    }

    @Test
    void flattenNullAndBlankYieldEmpty() {
        assertEquals("", TaskSteps.flattenForPrompt(null));
        assertEquals("", TaskSteps.flattenForPrompt(""));
    }
}

File: frontend/test/task-steps.test.ts
import { describe, it, expect } from 'vitest'
import { parseTaskSteps, serializeTaskSteps } from '~/utils/task-steps'

describe('parseTaskSteps (JCLAW-260 — mirrors backend TaskSteps.parse)', () => {
  it('plain text → single step', () => {
    expect(parseTaskSteps('Water the plants')).toEqual(['Water the plants'])
  })

  it('JSON array of strings → ordered steps', () => {
    expect(parseTaskSteps('["Fetch orders","Summarise","Post to Slack"]'))
      .toEqual(['Fetch orders', 'Summarise', 'Post to Slack'])
  })

  it('single-element array → one step', () => {
    expect(parseTaskSteps('["only step"]')).toEqual(['only step'])
  })

  it('null / blank → empty list', () => {
    expect(parseTaskSteps(null)).toEqual([])
    expect(parseTaskSteps('')).toEqual([])
    expect(parseTaskSteps('   ')).toEqual([])
  })

  it('non-string input → empty list', () => {
    expect(parseTaskSteps(undefined)).toEqual([])
    expect(parseTaskSteps(123)).toEqual([])
    expect(parseTaskSteps({ a: 1 })).toEqual([])
  })

  it('malformed array → single (raw) step', () => {
    expect(parseTaskSteps('[not valid json')).toEqual(['[not valid json'])
  })

  it('non-string array elements → single (raw) step', () => {
    expect(parseTaskSteps('["ok", 42, true]')).toEqual(['["ok", 42, true]'])
  })

  it('empty array → single (raw) step', () => {
    expect(parseTaskSteps('[]')).toEqual(['[]'])
  })
})

describe('serializeTaskSteps (JCLAW-22 — inverse of parseTaskSteps)', () => {
  it('multiple steps → JSON array string', () => {
    expect(serializeTaskSteps(['a', 'b', 'c'])).toBe('["a","b","c"]')
  })

  it('single step → plain text (not wrapped)', () => {
    expect(serializeTaskSteps(['just one'])).toBe('just one')
  })

  it('trims steps and drops blanks', () => {
    expect(serializeTaskSteps([' a ', '', '   ', 'b'])).toBe('["a","b"]')
  })

  it('all-empty → empty string', () => {
    expect(serializeTaskSteps([])).toBe('')
    expect(serializeTaskSteps(['', '  '])).toBe('')
  })

  it('round-trips with parseTaskSteps', () => {
    for (const steps of [['one'], ['a', 'b'], ['x', 'y', 'z']]) {
      expect(parseTaskSteps(serializeTaskSteps(steps))).toEqual(steps)
    }
  })
})
